src/shim.c

```diff
@@ -188,7 +188,7 @@ serialize_frame(struct frame *fr, ssize_t *total_size)
 	set_big_endian_32(msg + PAYLOAD_LEN_OFFSET, fr->header.payload_len);
 
 	if (fr->header.payload_len && fr->payload) {
-		strncpy((char *)msg + PAYLOAD_OFFSET, (const char*)fr->payload,
+		memcpy(msg + PAYLOAD_OFFSET, (const void*)fr->payload,
```

Comment:
const correctness would be to put const in the argument of this function, `const struct frame *fr`, signalling that we only read data from this pointer. I believe we should be able to get rid of this cast then.